src/util.h

```diff
@@ -35,7 +35,7 @@ using namespace std;
 NORETURN void Fatal(const char* msg, ...);
 
 // Have a generic fall-through for different versions of C/C++.
-#if defined(__cplusplus) && __cplusplus >= 201703L
+#if defined(__cplusplus) && __cplusplus >= 201703L || defined(__INTEL_COMPILER)
```

Comment:
Shouldn't the Intel compiler set `__cplusplus`?

Reply:
* Intel [defines](https://software.intel.com/en-us/cpp-compiler-developer-guide-and-reference-additional-predefined-macros) `__cplusplus=1` instead of a specific C++ revision level.
* Intel [fallthrough support](https://software.intel.com/en-us/articles/c17-features-supported-by-intel-c-compiler) is on all [currently supported versions](https://software.intel.com/en-us/articles/intel-parallel-studio-xe-supported-and-unsupported-product-versions) (2018 version expected to be dropped from support ~ March 2020)

Reply:
> Intel defines __cplusplus=1 instead of a specific C++ revision level.

Well, they aren't complying with the standard then. I don't want to include any workarounds for the Intel compiler (even if it's just a ifdef), sorry.

Reply:
@jhasse 

You want to support 10 plus year old versions of GCC, but you won't support a compiler released less than 2 years ago because of a single line change to an ifdef?

That's messed up.

Reply:
@jonesmz Can you try to be more friendly in your comments? You often don't add much to a discussion other than saying how incompetent I am (or the "ninja project" is) apparently. It hurts and doesn't help your case.

Reply:
If you feel that I consider you or the Ninja project in general to be incompetent, that is not the case, and I'm sad that you have that impression. I don't believe that I've ever said anything with the intention of implying that I think anyone here is incompetent, and would be happy to be educated on bad wording choices.

I am unhappy that the project has so many undocumented or under-documented polices, and I strongly disagree with several of the ones that have been documented, but that doesn't imply that I think anyone is incompetent. What I do think is that the project is not consistent with it's own stated policies and goals.

This particular PR is one example of the inconsistencies. Either the project should support ancient and/or questionably standards complaint compilers, or it shouldn't. The Intel compiler in question may not be standards compliant in this specific way, but at least the version of the compiler being discussed is new-ish and still under official support from the vendor.

Further, the cost of compatibility was only one line of code, in an existing #ifdef clause no less. (at the time this PR was still opened anyway, I see that @scivision has found a different way to implement support for the Intel compiler).

On the other hand, we have a version of getopt from 2011 bundled into the code and patched to support AIX and it's weird requirements, and we're using std::auto_ptr (Widely considered to be absolute garbage, and one of the few library features actually removed from C++, ever) all over the place because we still support C++03 because of an unknown list of compilers and platforms that no one will name.

I'm not saying that supporting older platforms is bad. I'm saying that if we're going to support older platforms, with all of the compromises that doing so requires, then rejecting compatibility with newer platforms is inconsistent. Either support everything that requires a little bit of tweaking, or don't support anything that requires non-standards compliant code. 

This project is fundamental infrastructure for a very large portion of world-wide computer technology, and is a tool that I use daily in my professional work. I have a strong interest in seeing it improved, and I've supplied code, or extensive code-review, on every issue that I have with the Ninja project, at no small time cost, in order to do so.

Reply:
> I don't believe that I've ever said anything with the intention of implying that I think anyone here is incompetent, and would be happy to be educated on bad wording choices.

Try to avoid rhetorical questions and "*shrug*".

> What I do think is that the project is not consistent with it's own stated policies and goals.

Most inconsistencies come from different people making their own independent decisions.

> On the other hand, we have a version of getopt from 2011 bundled into the code and patched to support AIX and it's weird requirements

AIX isn't the only reason for getopt, I think Windows among other platforms requires it, too.

> I'm not saying that supporting older platforms is bad. I'm saying that if we're going to support older platforms, with all of the compromises that doing so requires, then rejecting compatibility with newer platforms is inconsistent.

I wouldn't consider the Intel compiler a "platform". Furthermore we certainly *do* support newer platforms, e.g. we'll use unique_ptr instead of auto_ptr on newer compilers.

> [...] or don't support anything that requires non-standards compliant code.

Extreme rules like this hardly every work out. You could say "try to avoid supporting anything that requires non-standards compliant code whenever possible" or "minimize usage of non-standard compliant code".